book/src/rfcs/ffmul.md

```diff
@@ -0,0 +1,816 @@
+
+Remember we only need to prove the first $3\ell$ bits of $p - r$ are zero, since everything is $\mod 2^t$ and  $t = 3\ell$.  It may not be clear how this prefix witness approach, proves the $3\ell$ bits are indeed zero because within $u_0$ and $u_1$ there are bits that are nonzero.  The key observation is that these bits are too high for $\mod 2^t$.
+
+By making the prefix argument with $v_0$ and $v_1$ we are proving that $u_0$ is something prefixed with $2\ell$ zeros and that $u_1$ is something prefixed with $\ell$ zeros.  Any nonzero bits after $3\ell$ do not matter, since everything is $\mod 2^t$.
+
+All that remains is to range check $v_0$ and $v_1$
+> 7. Range check $v_0 \in [0, 3]$
+> 8. Range check $v_1 =\in [0, 2^{\ell + 3})$
+
+## Subtractions
+
+Now let's revisit our second problem, the possibility of borrows when subtracting the limbs of the remainder.  In order to constrain
+
+$$
+2^{2\ell}(p_2 - r_2) + 2^{\ell}(p_1 - r_1) + p_0 - r_0 = 0 \mod 2^t.
+$$
+
+we end up with some subtractions on the limbs.  As we previously mentioned, this a complexity we would like to avoid.  If our overall approach were to constrain this directly (i.e. if we had enough space in $\mathbb{F_n}$), then we would keep track of borrow bits $c_i$ for each limb $i$.  For example, we'd have constraints something like
+
+$$
+\begin{aligned}
+a_0 \cdot b_0 &= q_0 \cdot f'_0 + r_0 - c_0 \cdot 2^{2\ell + 1} \\
+a_1 \cdot b_1 &= q_1 \cdot f'_1 + r_1 - c_1 \cdot 2^{2\ell + 1} + c_0 \\
+a_2 \cdot b_2 &= q_2 \cdot f'_2 + r_2 + c_1
+\end{aligned}
+$$
+
+However, in our approach with $u_0$ and $u_1$ things are different.  We have
+
+$$
+\begin{aligned}
+u_0 &= p_0 + 2^{\ell}\cdot p_{10} - r_0 - 2^{\ell}\cdot r_1 \\
+&= 2^{\ell}(p_{10} - r_1) + p_0 - r_0
+\end{aligned}
+$$
+
+where $p_0 - r_0$ is $2\ell + 1$ bits and $2^{\ell}(p_{10} - r_1)$ is $2\ell$ bits.  Let's consider each possible underflow.
+
+Firstly, if $p_0 - r_0$ created an underflow we would borrow from $p_1 - r_1$ by adding $2^{2\ell + 1}$ to $p_0 - r_0$ and subtracting $1$ from $p_1 - r_1$.  However, recall that according to our definition of $u_0$, we already added to it the lowest $\ell$ bits of $p_1$ (i.e. $p_{10}$) and all of $r_1$.  So, we've already added the bits we would borrow.  Thus, no borrowing is needed here because it already implicitly accounted for.
+
+Secondly, if $p_{10} - r_1$ created an underflow, we would borrow from $p_{11}$ or $p_2 - r_2$.  This is the same as saying, we'd borrow from $p_{11} + p_2 - r_2 = u_1$.  Thus, we would add $2^{2\ell}$ to $p_{10} - r_1$ and subtract $1$ from $u_1$.  Thus, the constraint becomes
+
+$$
+u_0 = 2^{\ell}(p_{10} - r_1) + p_0 - r_0 - c_0 \cdot 2^{2\ell}
+$$
+
+For $u_1$ it's not possible for $r_2 > p_2$ and, therefore, the most significant borrow bit $c_1$ should always be zero.  Thus, $u_1$'s constraint is
+
+$$
+u_1 = p_{11} + p_2 - r_2 + c_0
```

Comment:
The `c` is a separate notion from the `v`'s.  The carry value cannot be in `v0` because part of `v0` is already included in the minuend `p0`.  I've updated the text and approach in this section to be more clear.

Reply:
> if there was a borrow, then the constraint about checking that the 2L least significant bits were zero wouldn't hold. Because for this to hold, the 2L LSB of p0 + p10 should be the same ones as the L of r0 concatenated to the L of r1.

I understand what you mean. If the constraint holds, then the $2\ell$ lowest bits of $p_0 + 2^{\ell} \cdot p_{10}$ minus the $2\ell$ bit $r_0 + 2^{\ell} \cdot r_1$ should be all `0`.  This implies that the $2\ell$ lowest bits of $p_0 + 2^{\ell} \cdot p_{10}$ are the same as $r_0 + 2^{\ell} \cdot r_1$.

I don't think what I'm suggesting would impact that.  Because when $u_0$ borrows from $u_1$ it borrows $2^{2\ell + 2}$, which doesn't change the $2\ell$ lowest bits of $p_0 + 2^{\ell} \cdot p_{10} - (r_0 + 2^{\ell} \cdot r_1)$.  It impacts the value of $v_0$'s highest bit.

Reply:
When unconstrained $p_0 + 2^{\ell} \cdot p_{10} - (r_0 + 2^{\ell} \cdot r_1)$ could require borrowing.  Fortunately, we have the constraint that the $2\ell$ least significant bits of $p_0 + 2^{\ell} \cdot p_{10} - (r_0 + 2^{\ell} \cdot r_1)$ are `0`, which means borrowing cannot happen.

Borrowing is prevented because when the the $2\ell$ least significant bits of the result are `0` it is not possible for the minuend to be less than the subtrahend.  We can prove this by contradiction.

Let

* $x = p_0 + 2^{\ell} \cdot p_{10}$
* $y = r_0 + 2^{\ell} \cdot r_1$
* the $2\ell$ least significant bits of $x - y$ be `0`

Suppose that borrowing occurs, that is, that $x < y$.  Recall that the length of $x$ is $2\ell + 2$ bits.  Therefore, since $x < y$ the top two bits of $x$ must be zero and so we have

$$
x - y = x_{2\ell} - y,
$$

where $x_{2\ell}$ denotes the $2\ell$ least significant bits of $x$.

Recall also that the length of $y$ is $2\ell$ bits.  So the result of this subtraction is $2\ell$ bits.   Since the $2\ell$ least significant bits of the subtraction are `0` this means that

$$
\begin{aligned}
x - y  &  = 0 \\
x &= y,
\end{aligned}
$$

which is a contradiction with $x < y$.